design/web-authn-authenticator.md

```diff
@@ -0,0 +1,429 @@
+# W3C Web Authentication - Implementation stages - Authenticator only
+
+* **Status**: Draft
+* **Author**: [tnorimat](https://github.com/tnorimat)
+* **JIRA**: [KEYCLOAK-9358](https://issues.jboss.org/browse/KEYCLOAK-9358), [KEYCLOAK-9359](https://issues.jboss.org/browse/KEYCLOAK-9359), [KEYCLOAK-9360](https://issues.jboss.org/browse/KEYCLOAK-9360)
+
+## Motivation
+
+As mentioned in [**W3C Web Authentication - Two-Factor**](https://github.com/keycloak/keycloak-community/blob/master/design/web-authn-two-factor.md), it is important to implement WebAuthn support to keycloak. This design document treats the following three issues mentioned in [**W3C Web Authentication - Two-Factor**](https://github.com/keycloak/keycloak-community/blob/master/design/web-authn-two-factor.md) Implementation stages - Authenticator only:
+
+1. [KEYCLOAK-9360](https://issues.jboss.org/browse/KEYCLOAK-9360) : Two factor authentication with W3C Web Authentication
+
+>  Acceptance criteria:
+>
+>    * Admin manually registers Web Authentication authenticator to replace OTP authenticator
+>    * Admin registers required action for user to register a security key
+>    * When user logs in again after registering the security key the user should be prompted to tap the security key after entering username and password
+
+2. [KEYCLOAK-9358](https://issues.jboss.org/browse/KEYCLOAK-9358) : Investigate W3C Web Authentication libraries
+
+>  The selected library should ideally support:
+>
+>    * MFA and passwordless
+>    * Username stored on device (no need to enter username when authenticating)
+>    * Attestation to be able to get details about authenticators and limit what authenticators should be supported by a realm
+>
+>  It should also have:
+>
+>    * A compatible open source license
+>    * A reasonable sized community
+>    * Good test coverage
+>    * Maven build (Graddle is ok, but not ideal)
+
+3. [KEYCLOAK-9359](https://issues.jboss.org/browse/KEYCLOAK-9359) : Investigate how to test Web Authentication
+
+## Relationship among other design documents
+***
+
+This design documents depends on the following two design documents:
+
+1. [**Application Initialted Actions**](https://github.com/keycloak/keycloak-community/blob/master/design/application-initiated-actions.md)
+
+It relates to how a user registers their public key credential on keycloak.
+
+Before realizing [**Application Initialted Actions**](https://github.com/keycloak/keycloak-community/blob/master/design/application-initiated-actions.md), an existing user having their user account can only register their public key credential in [User Account Service](https://www.keycloak.org/docs/latest/server_admin/index.html#_account-service).

```

Comment:
Application Initiated Actions will be used by account console as well, so until this is added it won't be possible to register public key credential except with a required action at login time.

Reply:
I see that point. I'll revise this design document including the following points:

* before realizing AIA, the user can register their own public key credential at their login time by Required Action.
* after realizing AIA, the user can also do it by AIA that User Account Service also use it.

Is that OK?

Reply:
+1